Firestore/core/src/firebase/firestore/model/field_path.cc

```diff
@@ -78,6 +81,56 @@ struct JoinEscaped {
 };
 }  // namespace
 
+FieldPath FieldPath::FromDotSeparatedString(absl::string_view path) {
```

Comment:
This method is mostly copied from FromServerFormat below, but it doesn't have the escaping and we throw invalid argument exceptions instead of asserting on errors.  There are a couple other approaches I could take:
1. Create a shared Parse() method that takes a `server_format` boolean parameter that changes its behavior.
2. Use std::stringstream to do the splitting (https://stackoverflow.com/a/236803), though this will presumably lose the special `\0` handling we have here (which may or may not be important??), and I'll have to do a second pass over the segments to check for empty ones.

If you have strong feelings, let me know.

Reply:
I think this function can be simplified. I think `FromServerFormat` contains quite a bit of complexity, and it's better to avoid that in this case, which is more straightforward. I'd suggest something like this (modulo comments):
```cpp
// Like before
 if (path.find_first_of("~*/[]") != absl::string_view::npos) {
    ThrowInvalidArgument(
        "Invalid field path (%s). Paths must not contain '~', '*', '/', '[', "
        "or ']'",
        path);
 }

std::vector<std::string> segments = absl::StrSplit(path, '.', [](absl::string_view segment) {
    // This is the predicate that decides whether to include the segment in the result.
    // However, rather than omitting the segment, we abort.
    if (segment.empty()) {
      ThrowInvalidArgument(
          "Invalid field path (%s). Paths must not be empty, begin with "
          "'.', end with '.', or contain '..'",
          path);
    }
    return true;
});
```

In general, Abseil has a few string manipulation utilities, which are always worth a look before doing a custom implementation.

Handling an embedded terminating null is _probably_ overkill. I'm not sure if it's supposed to be supported -- if not, maybe just do a `find_first_of` and throw if a `\0` is found?

Reply:
Oh, yes.  That's much better, thanks.  I was surprised no obvious string-splitting function came up in the SO answers, but I didn't think to specifically look in abseil.  I will next time.

And thanks for the code snippet! (you should feel free to just point me to the function next time rather than writing the code for me, though I certainly don't mind 😄).

I'll leave this unresolved in case Gil wants to chime in about the `\0` handling which I have no context on...

Reply:
The embedded null handling just preserves the behavior of the Objective-C client, and it's an artifact of the behavior of `NSString`.

I'm fine with either implementation, though I'm less partial to using esoteric features of `StrSplit` because they don't have analogues on the web or Android. For example, on Android, we split the string on dots and validate it in a separate pass:

https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/main/java/com/google/firebase/firestore/FieldPath.java#L61

Reply:
I'm leaving this as-is (with the slightly-clever StrSplit() implementation).  In general I'm not too concerned with porting differences within a method.  I think from a porting-perspective, it's much more problematic that we're doing API validation / parsing within a model:: class, but per our discussions, I think it's still the best pragmatic approach. I did add a couple PORTING NOTE comments though.